### ISSUE TITLE
[PYD-146] mypy plugin fails to resolve generics when used in inheritance 

### ISSUE THREAD

**Issue by sarahhenkens** [pydantic/pydantic#2613]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.8.1
            pydantic compiled: False
                 install path: /Users/sarah/projects/followmytv/hive/.venv/lib/python3.9/site-packages/pydantic
               python version: 3.9.1 (default, Jan 12 2021, 21:01:39)  [Clang 12.0.0 (clang-1200.0.31.1)]
                     platform: macOS-11.2.1-x86_64-i386-64bit
     optional deps. installed: ['devtools', 'dotenv', 'email-validator', 'typing-extensions']
```

## Description

I'm working on a generic Model which is considered an abstract base class, its only generic variable defines the type of one of the fields in this base class. Subclasses pass this type in as the generic value when they specify the base class as its superclass.

Data loading, exporting, and validation works as expected. Attribute access is getting warnings from mypy with the following error when attempting to use the property that is set via the generic:
```
 error: "T" has no attribute "foo"  [attr-defined]
```

### Example code

```py
from typing import Any, Dict, Generic, TypeVar

from pydantic import BaseModel
from pydantic.generics import GenericModel

Tbody = TypeVar("Tbody")


class Response(GenericModel, Generic[Tbody]):
    url: str
    body: Tbody


class JsonBody(BaseModel):
    raw: str
    data: Dict[str, Any]


class HtmlBody(BaseModel):
    raw: str
    doctype: str


class JsonResponse(Response[JsonBody]):
    pass


class HtmlResponse(Response[HtmlBody]):
    def custom_method(self) -> None:
        doctype = self.body.doctype. # Line 30
        print(f"self: {doctype}")


example = {"url": "foo.com", "body": {"raw": "..<html>..", "doctype": "html"}}

resp = HtmlResponse.parse_obj(example)
resp.custom_method()

doctype = resp.body.doctype # Line 39
print(f"main: {doctype}")
...
```

### Example's output
```
self: html
main: html
```

### Example's mypy results
```
generic.py:30: error: "Tbody" has no attribute "doctype"  [attr-defined]
generic.py:39: error: "Tbody" has no attribute "doctype"  [attr-defined]
```

### After I remove the pydantic mypy plugin
```
Success: no issues found in 1 source file
```

<sub>[PYD-146](https://linear.app/pydantic/issue/PYD-146/mypy-plugin-fails-to-resolve-generics-when-used-in-inheritance)</sub>

**Comment by mpkocher** [comment]
I believe I've run into a similar friction point with Pydantic + mypy. The workaround for mypy I used was:

```python
JsonResponse = Response[JsonBody]

_HtmlResponse = Response[HtmlBody]

class HtmlResponse(_HtmlResponse):
    def custom_method(self) -> None:
        doctype = self.body.doctype # Line 30
        print(f"self: {doctype}")
```

https://gist.github.com/mpkocher/0242865a0c5133a595dd3ae16a30b8dd

**Comment by wlf100220** [comment]
I have the same issue and I believe it is a bug.

**Comment by wlf100220** [comment]
> I believe I've run into a similar friction point with Pydantic + mypy. The workaround for mypy I used was:
> 
> ```python
> JsonResponse = Response[JsonBody]
> 
> _HtmlResponse = Response[HtmlBody]
> 
> class HtmlResponse(_HtmlResponse):
>     def custom_method(self) -> None:
>         doctype = self.body.doctype # Line 30
>         print(f"self: {doctype}")
> ```
> 
> https://gist.github.com/mpkocher/0242865a0c5133a595dd3ae16a30b8dd

Great temporal solution.
## In addition:
`init_forbid_extra = False` in mypy.ini should be set. Otherwise we can't add new field to `HtmlResponse`

**Comment by classiqRoi** [comment]
Just found a similar error in my project, happens on :
```
             pydantic version: 1.10.6
            pydantic compiled: True
                 install path: /Users/roigarf/Library/Caches/pypoetry/virtualenvs/backend-lACXRLDI-py3.8/lib/python3.8/site-packages/pydantic
               python version: 3.8.13 (default, Jan  8 2023, 12:08:28)  [Clang 14.0.0 (clang-1400.0.29.202)]
                     platform: macOS-13.1-arm64-arm-64bit
     optional deps. installed: ['dotenv', 'typing-extensions']
```
The solution also seems to work, but fixing this could help.
Is this going to be fixed soon?

**Comment by dmontagu** [comment]
While I believe this is still broken in pydantic v1, I'll note that this appears to be resolved in v2, with the appropriate changes.

```python
from typing import Any, Dict, Generic, TypeVar

from pydantic import BaseModel

Tbody = TypeVar("Tbody")


class Response(BaseModel, Generic[Tbody]):
    url: str
    body: Tbody


class JsonBody(BaseModel):
    raw: str
    data: Dict[str, Any]


class HtmlBody(BaseModel):
    raw: str
    doctype: str


class JsonResponse(Response[JsonBody]):
    pass


class HtmlResponse(Response[HtmlBody]):
    def custom_method(self) -> None:
        doctype = self.body.doctype  # Line 30
        print(f"self: {doctype}")


example = {"url": "foo.com", "body": {"raw": "..<html>..", "doctype": "html"}}

resp = HtmlResponse.model_validate(example)
resp.custom_method()

doctype = resp.body.doctype  # Line 39
reveal_type(resp.body)
# scratch.py:39: note: Revealed type is "scratch_56.HtmlBody"
reveal_type(JsonResponse.model_validate(example).body)
# scratch.py:40: note: Revealed type is "scratch_56.JsonBody"

print(f"main: {doctype}")
...
```

I'm open to reviewing a PR to the pydantic plugin to get this fixed in v1, but we are currently focused on getting the v2 release out so, as of now, don't intend to fix this prior to the v2 release.

**Comment by dmontagu** [comment]
It seems this is back to being broken on v2; I will investigate further once I have a chance to fix the issues with the plugin, hopefully soon.

**Comment by dmontagu** [comment]
This is now fixed in v2.0 (which has just been released 🚀). Fixing it involved a significant update to the plugin, but if it is important to backport a fix to the v1 branch, it may be possible to do that without _too_ much pain. However, without significant interest specifically in supporting this in v1.10, I'm not currently planning to backport the fix (due to the complexity of the overhaul to the plugin).

However, if anyone wants to take a stab at backporting the fix, I would review a PR.
